### ISSUE TITLE
spring-boot-autoconfigure + spring-test: java.io.FileNotFoundException: class path resource [....class] cannot be opened because it does not exist

### ISSUE THREAD

**Issue by hohwille** [spring-projects/spring-boot#18868]
When running JUnit Tests with something like `@SpringBootTest(classes = SpringBootApplication.class, webEnvironment = WebEnvironment.NONE)` I get errors like:
` java.io.FileNotFoundException: class path resource [org/springframework/session/hazelcast/config/annotation/web/http/HazelcastHttpSessionConfiguration.class] cannot be opened because it does not exist`

What seems to origninate from here:
https://github.com/spring-projects/spring-boot/blob/f78c9574d8058a3ac8203ce3d649a59b96266581/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/session/HazelcastSessionConfiguration.java#L32

The entire idea of spring-boot autoconfigurations was according to my understanding that they are triggered if a specific class is found on classpath and then they can do some integration magic otherwise they should do nothing. However, if spring-boot (or more precisely spring-test) fails if that dependency is not found this renders spring-boot pretty much useless as `spring-boot-autoconfigure` has so many optional dependencies including commercial products that I would all have to add to my project.

Environment:
Java 11.0.4, spring-boot 2.1.6.RELEASE (also tested with e.g. 2.1.9.RELEASE), spring-test 5.1.8.RELEASE.
Happens both in Eclipse as well as in maven via bash (maven 3.6.1).
Fails with JUnit (tested with JUnit4 and JUnit5), but works if I run `SpringBootApp` via `main` method as regular Java app.

So my assumption is that spring-test for some reason broke the autoconfig mechanism.
If you want me to do more specific tests I am happy to assist.

In case you want to reproduce, create a new app like this:
```
mvn -DarchetypeVersion=3.1.0 -DarchetypeGroupId=com.devonfw.java.templates -DarchetypeArtifactId=devon4j-template-server archetype:generate -DgroupId=com.example.application -DartifactId=sampleapp -Dversion=1.0.0-SNAPSHOT -Dpackage=com.devonfw.application.sampleapp
```
And in the new app create a simple test like this:
```
@SpringBootTest(classes = SpringBootApplication.class, webEnvironment = WebEnvironment.NONE)
public class SpringBootErrorTest extends ComponentTest {

  @Test
  public void test() { }
}
```
For more recent versions and Junit5 use `3.2.0` as `archetypeVersion` instead.
Same error also with `WebEnvironment.MOCK`.

Full log:
```
[D: 2019-11-04 19:47:13,817] [P: INFO ] [C: ] [T: main] [L: org.springframework.boot.test.context.SpringBootTestContextBootstrapper] - [M: Neither @ContextConfiguration nor @ContextHierarchy found for test class [org.example.app.dummy.common.api.ComponentATest], using SpringBootContextLoader]  
[D: 2019-11-04 19:47:13,840] [P: INFO ] [C: ] [T: main] [L: org.springframework.test.context.support.AbstractContextLoader] - [M: Could not detect default resource locations for test class [org.example.app.dummy.common.api.ComponentATest]: no resource found for suffixes {-context.xml, Context.groovy}.]  
[D: 2019-11-04 19:47:14,187] [P: INFO ] [C: ] [T: main] [L: org.springframework.boot.test.context.SpringBootTestContextBootstrapper] - [M: Using TestExecutionListeners: [org.springframework.test.context.transaction.TransactionalTestExecutionListener@1dd6d4b7, org.springframework.test.context.support.DependencyInjectionTestExecutionListener@56e8b606]]  

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::        (v2.1.6.RELEASE)

[D: 2019-11-04 19:47:14,738] [P: INFO ] [C: ] [T: main] [L: org.example.app.dummy.common.api.ComponentATest] - [M: Starting ComponentATest on Jorgs-MacBook-Pro.local with PID 18946 (started by hohwille in /projects/devonfw-training/workspaces/main/app/core)]  
[D: 2019-11-04 19:47:14,739] [P: DEBUG] [C: ] [T: main] [L: org.example.app.dummy.common.api.ComponentATest] - [M: Running with Spring Boot v2.1.6.RELEASE, Spring v5.1.8.RELEASE]  
[D: 2019-11-04 19:47:14,739] [P: INFO ] [C: ] [T: main] [L: org.example.app.dummy.common.api.ComponentATest] - [M: The following profiles are active: junit]  
[D: 2019-11-04 19:58:50,351] [P: WARN ] [C: ] [T: main] [L: org.springframework.context.annotation.AnnotationConfigApplicationContext] - [M: Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanDefinitionStoreException: Failed to parse configuration class [org.springframework.boot.autoconfigure.SpringBootApplication]; nested exception is java.io.FileNotFoundException: class path resource [org/springframework/session/hazelcast/config/annotation/web/http/HazelcastHttpSessionConfiguration.class] cannot be opened because it does not exist]  
[D: 2019-11-04 19:58:50,368] [P: INFO ] [C: ] [T: main] [L: org.springframework.boot.autoconfigure.logging.ConditionEvaluationReportLoggingListener] - [M: 

Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.]  
[D: 2019-11-04 19:58:50,374] [P: ERROR] [C: ] [T: main] [L: org.springframework.boot.SpringApplication] - [M: Application run failed]  
org.springframework.beans.factory.BeanDefinitionStoreException: Failed to parse configuration class [org.springframework.boot.autoconfigure.SpringBootApplication]; nested exception is java.io.FileNotFoundException: class path resource [org/springframework/session/hazelcast/config/annotation/web/http/HazelcastHttpSessionConfiguration.class] cannot be opened because it does not exist
	at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:181)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:315)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:232)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:275)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:95)
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:705)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:531)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:742)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:389)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:311)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:119)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:99)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:117)
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:108)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.injectDependencies(DependencyInjectionTestExecutionListener.java:118)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.prepareTestInstance(DependencyInjectionTestExecutionListener.java:83)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:246)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:227)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:289)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:291)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:246)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:763)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:463)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:209)
Caused by: java.io.FileNotFoundException: class path resource [org/springframework/session/hazelcast/config/annotation/web/http/HazelcastHttpSessionConfiguration.class] cannot be opened because it does not exist
	at org.springframework.core.io.ClassPathResource.getInputStream(ClassPathResource.java:180)
	at org.springframework.core.type.classreading.SimpleMetadataReader.<init>(SimpleMetadataReader.java:51)
	at org.springframework.core.type.classreading.SimpleMetadataReaderFactory.getMetadataReader(SimpleMetadataReaderFactory.java:103)
	at org.springframework.boot.type.classreading.ConcurrentReferenceCachingMetadataReaderFactory.createMetadataReader(ConcurrentReferenceCachingMetadataReaderFactory.java:86)
	at org.springframework.boot.type.classreading.ConcurrentReferenceCachingMetadataReaderFactory.getMetadataReader(ConcurrentReferenceCachingMetadataReaderFactory.java:73)
	at org.springframework.core.type.classreading.SimpleMetadataReaderFactory.getMetadataReader(SimpleMetadataReaderFactory.java:81)
	at org.springframework.context.annotation.ConfigurationClassParser.asSourceClass(ConfigurationClassParser.java:682)
	at org.springframework.context.annotation.ConfigurationClassParser$SourceClass.getSuperClass(ConfigurationClassParser.java:995)
	at org.springframework.context.annotation.ConfigurationClassParser.doProcessConfigurationClass(ConfigurationClassParser.java:332)
	at org.springframework.context.annotation.ConfigurationClassParser.processConfigurationClass(ConfigurationClassParser.java:242)
	at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:191)
	at org.springframework.context.annotation.ConfigurationClassParser.doProcessConfigurationClass(ConfigurationClassParser.java:295)
	at org.springframework.context.annotation.ConfigurationClassParser.processConfigurationClass(ConfigurationClassParser.java:242)
	at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:199)
	at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:167)
	... 37 common frames omitted
[D: 2019-11-04 19:58:50,376] [P: ERROR] [C: ] [T: main] [L: org.springframework.test.context.TestContextManager] - [M: Caught exception while allowing TestExecutionListener [org.springframework.test.context.support.DependencyInjectionTestExecutionListener@56e8b606] to prepare test instance [org.example.app.dummy.common.api.ComponentATest@3dc95b8b]]  
java.lang.IllegalStateException: Failed to load ApplicationContext
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:125)
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:108)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.injectDependencies(DependencyInjectionTestExecutionListener.java:118)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.prepareTestInstance(DependencyInjectionTestExecutionListener.java:83)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:246)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:227)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:289)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:291)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:246)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:763)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:463)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:209)
Caused by: org.springframework.beans.factory.BeanDefinitionStoreException: Failed to parse configuration class [org.springframework.boot.autoconfigure.SpringBootApplication]; nested exception is java.io.FileNotFoundException: class path resource [org/springframework/session/hazelcast/config/annotation/web/http/HazelcastHttpSessionConfiguration.class] cannot be opened because it does not exist
	at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:181)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:315)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:232)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:275)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:95)
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:705)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:531)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:742)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:389)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:311)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:119)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:99)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:117)
	... 25 common frames omitted
Caused by: java.io.FileNotFoundException: class path resource [org/springframework/session/hazelcast/config/annotation/web/http/HazelcastHttpSessionConfiguration.class] cannot be opened because it does not exist
	at org.springframework.core.io.ClassPathResource.getInputStream(ClassPathResource.java:180)
	at org.springframework.core.type.classreading.SimpleMetadataReader.<init>(SimpleMetadataReader.java:51)
	at org.springframework.core.type.classreading.SimpleMetadataReaderFactory.getMetadataReader(SimpleMetadataReaderFactory.java:103)
	at org.springframework.boot.type.classreading.ConcurrentReferenceCachingMetadataReaderFactory.createMetadataReader(ConcurrentReferenceCachingMetadataReaderFactory.java:86)
	at org.springframework.boot.type.classreading.ConcurrentReferenceCachingMetadataReaderFactory.getMetadataReader(ConcurrentReferenceCachingMetadataReaderFactory.java:73)
	at org.springframework.core.type.classreading.SimpleMetadataReaderFactory.getMetadataReader(SimpleMetadataReaderFactory.java:81)
	at org.springframework.context.annotation.ConfigurationClassParser.asSourceClass(ConfigurationClassParser.java:682)
	at org.springframework.context.annotation.ConfigurationClassParser$SourceClass.getSuperClass(ConfigurationClassParser.java:995)
	at org.springframework.context.annotation.ConfigurationClassParser.doProcessConfigurationClass(ConfigurationClassParser.java:332)
	at org.springframework.context.annotation.ConfigurationClassParser.processConfigurationClass(ConfigurationClassParser.java:242)
	at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:191)
	at org.springframework.context.annotation.ConfigurationClassParser.doProcessConfigurationClass(ConfigurationClassParser.java:295)
	at org.springframework.context.annotation.ConfigurationClassParser.processConfigurationClass(ConfigurationClassParser.java:242)
	at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:199)
	at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:167)
	... 37 common frames omitted

```

**Comment by wilkinsona** [comment]
Thanks for the snippets and the steps to reproduce. The problem is caused by your use of `@SpringBootTest`:

```java
@SpringBootTest(classes = SpringBootApplication.class, webEnvironment = WebEnvironment.NONE)
public class SpringBootErrorTest extends ComponentTest {

  @Test
  public void test() { }
  
}
```

This is telling the test framework to use `@SpringBootApplication` as a source class for your application. It is annotated with `@ComponentScan` so it enables component scanning for the `org.springframework.boot.autoconfigure` package which results in the Spring Session auto-configuration being activated unconditionally.

To fix the problem, the `classes` attribute should point to your application class that is annotated with `@SpringBootApplication` (or `@SpringBootConfiguration`). In the case of the sample, the class is `com.devonfw.application.sampleapp.SpringBootApp`:

```java
package com.devonfw.application.sampleapp;

import org.junit.Test;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;

import com.devonfw.module.test.common.base.ComponentTest;

@SpringBootTest(classes = SpringBootApp.class, webEnvironment = WebEnvironment.NONE)
public class SpringBootErrorTest extends ComponentTest {

  @Test
  public void test() { }
  
}
```

**Comment by hohwille** [comment]
I did some more analysis on this. It seems to only occure in and edge-case. I tried to isolate the problem and was not able to reproduce it easily in a vanilla environment. So I will dig deeper till I found the real root casue. Until then, feel free to keep this as low prio or even close until I come back with the full details to reproduce the bug. I assume it will still be valuable to understand what went wrong here for spring-boot. However, do not put any effort into this before I come up with final details...
